File: tests/layer1_filters/unit/TestFirwin.m
classdef TestFirwin < matlab.unittest.TestCase
%TESTFIRWIN  Unit tests for firwin, freqwin, freqwavelet.
%
%   firwin(name, M)  — symmetric FIR window of length M
%   freqwin(name, L, bw)  — bandpass frequency window, peak-normalised
%   freqwavelet(name, L)  — mother wavelet in frequency domain, peak-normalised


    methods (TestClassSetup)
        function setupClass(tc)
            sharedDir = fullfile(fileparts(mfilename('fullpath')), '..', '..', 'shared');
            addpath(sharedDir);
            setup_filterbank_paths();
        end
    end

    % ── firwin: output length ────────────────────────────────────────────────
    methods (Test)

        function testFirwinOutputLength(tc)
            for M = [16, 32, 64, 128]
                w = firwin('hann', M);
                tc.verifyEqual(numel(w), M, ...
                    sprintf('firwin hann: expected length %d', M));
            end
        end

        function testFirwinOutputLengthOdd(tc)
            for M = [15, 33, 65]
                w = firwin('hann', M);
                tc.verifyEqual(numel(w), M, ...
                    sprintf('firwin hann odd: expected length %d', M));
            end
        end

        function testFirwinSineLength(tc)
            w = firwin('sine', 64);
            tc.verifyEqual(numel(w), 64);
        end

        function testFirwinRectLength(tc)
            w = firwin('rect', 32);
            tc.verifyEqual(numel(w), 32);
        end

    end

    % ── firwin: symmetry (whole-point even) ──────────────────────────────────
    methods (Test)

        function testFirwinHannWPESymmetry(tc)
            % w(k) = w(M+2-k) for k = 2..M  (whole-point even)
            M = 64;
            w = firwin('hann', M);
            tc.verifyEqual(w(2:end), flipud(w(2:end)), 'AbsTol', 1e-14, ...
                'hann: not whole-point even symmetric');
        end

        function testFirwinSineWPESymmetry(tc)
            M = 64;
            w = firwin('sine', M);
            tc.verifyEqual(w(2:end), flipud(w(2:end)), 'AbsTol', 1e-14, ...
                'sine: not whole-point even symmetric');
        end

        function testFirwinRectWPESymmetry(tc)
            M = 64;
            w = firwin('rect', M);
            tc.verifyEqual(w(2:end), flipud(w(2:end)), 'AbsTol', 1e-14, ...
                'rect: not whole-point even symmetric');
        end

        function testFirwinTriaWPESymmetry(tc)
            M = 64;
            w = firwin('tria', M);
            tc.verifyEqual(w(2:end), flipud(w(2:end)), 'AbsTol', 1e-14, ...
                'tria: not whole-point even symmetric');
        end

    end

    % ── firwin: non-negativity ───────────────────────────────────────────────
    methods (Test)

        function testFirwinHannNonNegative(tc)
            w = firwin('hann', 64);
            tc.verifyGreaterThanOrEqual(min(w), -1e-14, ...
                'hann window should be non-negative');
        end

        function testFirwinSineNonNegative(tc)
            w = firwin('sine', 64);
            tc.verifyGreaterThanOrEqual(min(w), -1e-14);
        end

        function testFirwinRectIsOnesVector(tc)
            % For even M, the sampling grid includes x = -M/2/M = -0.5, and
            % abs(-0.5) < 0.5 is false, so that one element is 0 (the docstring
            % calls it "(Almost) rectangular").  Use odd M to get a true all-ones
            % vector: odd grids are symmetric around 0 with |x| < 0.5 for all pts.
            M = 33;
            w = firwin('rect', M);
            tc.verifyEqual(w, ones(M, 1), 'AbsTol', 1e-14, ...
                'rect window (odd M) should be all-ones');
        end

    end

    % ── firwin: partition of unity ───────────────────────────────────────────
    methods (Test)

        function testFirwinHannPU(tc)
            % hann: w + fftshift(w) = ones  (for even M)
            M = 64;
            w = firwin('hann', M);
            tc.verifyEqual(w + fftshift(w), ones(M, 1), 'AbsTol', 1e-14, ...
                'hann: partition of unity w + fftshift(w) = 1 failed');
        end

        function testFirwinHannPUMultipleLengths(tc)
            for M = [32, 64, 128, 256]
                w = firwin('hann', M);
                tc.verifyEqual(w + fftshift(w), ones(M, 1), 'AbsTol', 1e-14, ...
                    sprintf('hann PU failed for M=%d', M));
            end
        end

        function testFirwinSineTightFrame(tc)
            % sine = sqrt(hann): sine.^2 + fftshift(sine.^2) = ones  (tight frame)
            M = 64;
            w = firwin('sine', M);
            tc.verifyEqual(w.^2 + fftshift(w.^2), ones(M, 1), 'AbsTol', 1e-14, ...
                'sine: tight frame condition sine^2 + fftshift(sine^2) = 1 failed');
        end

        function testFirwinSineTightFrameMultipleLengths(tc)
            for M = [32, 64, 128]
                w = firwin('sine', M);
                tc.verifyEqual(w.^2 + fftshift(w.^2), ones(M, 1), 'AbsTol', 1e-14, ...
                    sprintf('sine tight frame failed for M=%d', M));
            end
        end

    end

    % ── firwin: peak at DC (index 1) ─────────────────────────────────────────
    methods (Test)

        function testFirwinHannPeakAtDC(tc)
            % firwin returns zero-phase window: max at index 1
            w = firwin('hann', 64);
            [~, idx] = max(w);
            tc.verifyEqual(idx, 1, 'hann: peak should be at index 1 (DC)');
        end

        function testFirwinSinePeakAtDC(tc)
            w = firwin('sine', 64);
            [~, idx] = max(w);
            tc.verifyEqual(idx, 1, 'sine: peak should be at index 1 (DC)');
        end

    end

    % ── freqwin: output ──────────────────────────────────────────────────────
    methods (Test)

        function testFreqwinOutputLength(tc)
            L  = 128;
            bw = 0.1;
            w  = freqwin('gauss', L, bw);
            tc.verifyEqual(numel(w), L, 'freqwin gauss: wrong output length');
        end

        function testFreqwinPeakNormalized(tc)
            % freqwin default: peak-normalized, max|w| = 1
            L  = 256;
            bw = 0.05;
            w  = freqwin('gauss', L, bw);
            tc.verifyEqual(max(abs(w)), 1, 'AbsTol', 1e-12, ...
                'freqwin: should be peak-normalized');
        end

        function testFreqwinGammatoneLength(tc)
            w = freqwin('gammatone', 128, 0.1);
            tc.verifyEqual(numel(w), 128);
        end

        function testFreqwinButterworthLength(tc)
            w = freqwin('butterworth', 128, 0.1);
            tc.verifyEqual(numel(w), 128);
        end

        function testFreqwinNonNegative(tc)
            % Frequency-domain windows are non-negative (they are magnitude shapes)
            w = freqwin('gauss', 128, 0.1);
            tc.verifyGreaterThanOrEqual(min(w), -1e-12, ...
                'freqwin gauss: should be non-negative');
        end

    end

    % ── freqwavelet: output ──────────────────────────────────────────────────
    methods (Test)

        function testFreqwaveletOutputLength(tc)
            L = 256;
            H = freqwavelet('cauchy', L);
            tc.verifyEqual(numel(H), L, 'freqwavelet: wrong output length');
        end

        function testFreqwaveletPeakNormalized(tc)
            % freqwavelet defaults to 'null' normalization (no scaling).
            % Pass 'peak' explicitly to obtain max|H| = 1.
            H = freqwavelet('cauchy', 256, 'peak');
            tc.verifyEqual(max(abs(H)), 1, 'AbsTol', 1e-12, ...
                'freqwavelet cauchy with peak norm: max|H| should be 1');
        end

        function testFreqwaveletMorletLength(tc)
            H = freqwavelet('morlet', 256);
            tc.verifyEqual(numel(H), 256);
        end

        function testFreqwaveletMorseLength(tc)
            H = freqwavelet('morse', 256);
            tc.verifyEqual(numel(H), 256);
        end

        function testFreqwaveletFiniteValues(tc)
            % No NaN or Inf for standard parameters
            for name = {'cauchy', 'morlet', 'morse'}
                H = freqwavelet(name{1}, 256);
                tc.verifyTrue(all(isfinite(H)), ...
                    sprintf('freqwavelet %s: non-finite values', name{1}));
            end
        end

    end

end

File: tests/layer1_filters/unit/test_firwin.py
"""
test_firwin.py
==============
Python port of:
    layer1_filters/unit/TestFirwin.m

Covers: firwin, freqwin, freqwavelet

LTFAT frequency convention
---------------------------
Windows use a whole-point even (WPE) layout: peak at index 0, so that
w[k] = w[-k mod M] (i.e., w[1:] == w[1:][::-1]).

firwin 'hann' partition of unity: w + np.roll(fftshift(w))  — the Python
equivalent of MATLAB's fftshift is np.fft.fftshift.
"""

from __future__ import annotations

import pytest

import numpy as np
from conftest import firwin_ref

# ---------------------------------------------------------------------------
# firwin – reference tests (no impl required)
# ---------------------------------------------------------------------------

class TestFirwinReference:
    """
    Mathematical properties verified using firwin_ref.
    MATLAB counterpart: TestFirwin.
    """

    # ── output length ────────────────────────────────────────────────────────

    @pytest.mark.parametrize("M", [16, 32, 64, 128])
    def test_hann_output_length(self, M):
        assert len(firwin_ref("hann", M)) == M

    @pytest.mark.parametrize("M", [15, 33, 65])
    def test_hann_output_length_odd(self, M):
        assert len(firwin_ref("hann", M)) == M

    def test_sine_length(self):
        assert len(firwin_ref("sine", 64)) == 64

    def test_rect_length(self):
        assert len(firwin_ref("rect", 32)) == 32

    # ── WPE symmetry: w[1:] == w[1:][::-1] ───────────────────────────────────

    @pytest.mark.parametrize("name", ["hann", "sine", "rect", "tria"])
    @pytest.mark.parametrize("M", [18, 19, 20, 21, 64])
    def test_wpe_symmetry(self, name, M):
        w = firwin_ref(name, M)
        np.testing.assert_allclose(w[1:], w[1:][::-1], atol=1e-14,
                                   err_msg=f"{name} M={M}: not WPE symmetric")

    # ── non-negativity ────────────────────────────────────────────────────────

    def test_hann_non_negative(self):
        w = firwin_ref("hann", 64)
        assert np.min(w) >= -1e-14

    def test_sine_non_negative(self):
        w = firwin_ref("sine", 64)
        assert np.min(w) >= -1e-14

    def test_rect_odd_is_all_ones(self):
        """firwin('rect', odd M): all elements are 1."""
        w = firwin_ref("rect", 33)
        np.testing.assert_allclose(w, np.ones(33), atol=1e-14)

    # ── partition of unity: w + fftshift(w) = 1 ──────────────────────────────

    @pytest.mark.parametrize("M", [32, 64, 128, 256])
    def test_hann_pu(self, M):
        w   = firwin_ref("hann", M)
        pu  = w + np.fft.fftshift(w)
        np.testing.assert_allclose(pu, np.ones(M), atol=1e-14,
                                   err_msg=f"hann PU failed for M={M}")

    @pytest.mark.parametrize("M", [32, 64, 128, 256])
    def test_tria_pu(self, M):
        w   = firwin_ref("tria", M)
        pu  = w + np.fft.fftshift(w)
        np.testing.assert_allclose(pu, np.ones(M), atol=1e-14,
                                   err_msg=f"tria PU failed for M={M}")

    # ── sine tight frame: w^2 + fftshift(w^2) = 1 ────────────────────────────

    @pytest.mark.parametrize("M", [32, 64, 128])
    def test_sine_tight_frame(self, M):
        w   = firwin_ref("sine", M)
        tf  = w ** 2 + np.fft.fftshift(w ** 2)
        np.testing.assert_allclose(tf, np.ones(M), atol=1e-14,
                                   err_msg=f"sine tight frame failed for M={M}")

    def test_sine_is_sqrt_hann(self):
        """sine^2 == hann point-wise."""
        M    = 64
        sine = firwin_ref("sine", M)
        hann = firwin_ref("hann", M)
        np.testing.assert_allclose(sine ** 2, hann, atol=1e-14)

    # ── peak at index 0 ───────────────────────────────────────────────────────

    @pytest.mark.parametrize("name", ["hann", "sine", "rect", "tria"])
    def test_peak_at_dc(self, name):
        w = firwin_ref(name, 64)
        assert np.argmax(w) == 0, f"{name}: peak should be at index 0"

    @pytest.mark.parametrize("name", ["hann", "sine", "rect", "tria"])
    def test_peak_value_one(self, name):
        """All windows should peak at 1 (WPE convention)."""
        w = firwin_ref(name, 64)
        np.testing.assert_allclose(w[0], 1.0, atol=1e-14,
                                   err_msg=f"{name}: peak should be 1")


# ---------------------------------------------------------------------------
# firwin – implementation tests
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFirwinImpl:
    """Verify cool_frames.layer1.firwin matches reference."""

    @pytest.mark.parametrize("name,M", [
        ("hann", 32), ("hann", 64), ("sine", 64), ("rect", 33), ("tria", 64)
    ])
    def test_matches_reference(self, needs_impl, name, M):
        from cool_frames.filters import firwin  # type: ignore

        np.testing.assert_allclose(firwin(name, M), firwin_ref(name, M), atol=1e-12,
                                   err_msg=f"{name}, M={M}")

    # -- Broader parametric tests (ported from ttest_firwin.m) --

    @pytest.mark.parametrize("name", [
        "hann", "hamming", "blackman", "rect", "tria", "nuttall",
    ])
    @pytest.mark.parametrize("M", [18, 19, 20, 21])
    def test_wpe_symmetry_all_windows(self, needs_impl, name, M):
        """WPE symmetry: w[1:] == w[1:][::-1] for all windows and lengths."""
        from cool_frames.filters import firwin  # type: ignore
        w = firwin(name, M, norm="inf")
        np.testing.assert_allclose(w[1:], w[1:][::-1], atol=1e-14,
                                   err_msg=f"{name} M={M}: not WPE symmetric")

    @pytest.mark.parametrize("name", [
        "hann", "hamming", "blackman", "rect", "tria", "nuttall",
    ])
    def test_peak_at_bin1_all_windows(self, needs_impl, name):
        """Peak should be at index 0 (DFT ordering) for all windows."""
        from cool_frames.filters import firwin  # type: ignore
        w = firwin(name, 64, norm="inf")
        np.testing.assert_allclose(w[0], 1.0, atol=1e-10,
                                   err_msg=f"{name}: peak should be 1")


# ---------------------------------------------------------------------------
# freqwin – implementation tests only
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFreqwinImpl:
    """
    MATLAB counterpart: TestFirwin (freqwin section).
    """

    def test_gauss_output_length(self, needs_impl):
        from cool_frames.filters import freqwin  # type: ignore
        assert len(freqwin("gauss", 128, 0.1)) == 128

    def test_gauss_peak_normalized(self, needs_impl):
        from cool_frames.filters import freqwin  # type: ignore
        w = freqwin("gauss", 256, 0.05)
        assert max(abs(w)) == pytest.approx(1.0, abs=1e-12)

    def test_gammatone_length(self, needs_impl):
        from cool_frames.filters import freqwin  # type: ignore
        assert len(freqwin("gammatone", 128, 0.1)) == 128

    def test_butterworth_length(self, needs_impl):
        from cool_frames.filters import freqwin  # type: ignore
        assert len(freqwin("butterworth", 128, 0.1)) == 128

    def test_non_negative(self, needs_impl):
        from cool_frames.filters import freqwin  # type: ignore
        w = freqwin("gauss", 128, 0.1)
        assert np.min(w) >= -1e-12


# ---------------------------------------------------------------------------
# freqwavelet – implementation tests only
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFreqwaveletImpl:
    """
    MATLAB counterpart: TestFirwin (freqwavelet section).
    """

    def test_cauchy_output_length(self, needs_impl):
        from cool_frames.filters import freqwavelet  # type: ignore
        H, _ = freqwavelet("cauchy", 256)
        assert len(H) == 256

    def test_cauchy_peak_normalized(self, needs_impl):
        from cool_frames.filters import freqwavelet  # type: ignore
        H, _ = freqwavelet("cauchy", 256, "peak")
        assert max(abs(H)) == pytest.approx(1.0, abs=1e-12)

    def test_morlet_length(self, needs_impl):
        from cool_frames.filters import freqwavelet  # type: ignore
        H, _ = freqwavelet("morlet", 256)
        assert len(H) == 256

    def test_morse_length(self, needs_impl):
        from cool_frames.filters import freqwavelet  # type: ignore
        H, _ = freqwavelet("morse", 256)
        assert len(H) == 256

    @pytest.mark.parametrize("name", ["cauchy", "morlet", "morse"])
    def test_finite_values(self, needs_impl, name):
        from cool_frames.filters import freqwavelet  # type: ignore
        H, _ = freqwavelet(name, 256)
        assert np.all(np.isfinite(H)), f"freqwavelet {name}: non-finite values"
